File: session-3/python/service_validation.py
#!/usr/bin/env python3
"""
Service Validation — Verifie la coherence des donnees de capteurs.

Ce service consomme les messages AMQP depuis l'exchange amq.topic (binding maison.#),
verifie que les valeurs sont dans les plages acceptables, puis :
- Si valide : republier sur l'exchange "validated" avec la meme routing key
- Si invalide : rejeter le message (nack, sans requeue)

Utilise le manual ack pour garantir le traitement de chaque message.
"""

import json
import time

import pika

# Configuration AMQP
AMQP_HOST = "localhost"
AMQP_USER = "guest"
AMQP_PASS = "guest"

# Plages valides pour chaque type de capteur
PLAGES_VALIDES = {
    "temperature": (-20, 60),
    "humidite": (0, 100),
    "fumee": (0, 100),
}


def connecter_amqp():
    """
    Se connecte au broker AMQP avec logique de retry.
    Tente 10 fois avec un delai de 2 secondes entre chaque tentative.
    """
    credentials = pika.PlainCredentials(AMQP_USER, AMQP_PASS)
    parametres = pika.ConnectionParameters(
        host=AMQP_HOST,
        credentials=credentials
    )

    for tentative in range(1, 11):
        try:
            connexion = pika.BlockingConnection(parametres)
            print(f"[AMQP] Connecte a {AMQP_HOST}")
            return connexion
        except Exception as e:
            print(f"[AMQP] Tentative {tentative}/10 echouee : {e}")
            if tentative < 10:
                time.sleep(2)

    print("[AMQP] Impossible de se connecter apres 10 tentatives.")
    raise ConnectionError("Connexion AMQP impossible")


def valider_message(donnees):
    """
    Verifie si les donnees du capteur sont dans la plage valide.
    Retourne (True, None) si valide, (False, raison) si invalide.
    """
    type_capteur = donnees.get("sensor")
    valeur = donnees.get("value")

    if type_capteur == "mouvement":
        # Le mouvement doit etre un booleen
        if isinstance(valeur, bool):
            return True, None
        return False, f"doit etre un booleen, recu: {valeur}"

    if type_capteur in PLAGES_VALIDES:
        minimum, maximum = PLAGES_VALIDES[type_capteur]
        if minimum <= valeur <= maximum:
            return True, None
        return False, f"hors plage {minimum} a {maximum}"

    # Type de capteur inconnu : on considere invalide
    return False, f"type de capteur inconnu: {type_capteur}"


def callback(channel, method, properties, body):
    """
    Callback appele pour chaque message recu.
    Valide les donnees et les transmet ou les rejette.
    """
    try:
        donnees = json.loads(body)
    except json.JSONDecodeError:
        print(f"[INVALIDE] Message non-JSON recu: {body}")
        channel.basic_nack(delivery_tag=method.delivery_tag, requeue=False)
        return

    type_capteur = donnees.get("sensor", "inconnu")
    piece = donnees.get("room", "inconnue")
    valeur = donnees.get("value")

    # Valider les donnees
    est_valide, raison = valider_message(donnees)

    if est_valide:
        # Republier sur l'exchange "validated" avec la meme routing key
        channel.basic_publish(
            exchange="validated",
            routing_key=method.routing_key,
            body=body
        )
        print(f"[VALIDE]   {type_capteur} {piece}: {valeur}")
        channel.basic_ack(delivery_tag=method.delivery_tag)
    else:
        # Rejeter le message sans requeue
        unite = ""
        if type_capteur == "temperature":
            unite = "°C"
        elif type_capteur == "humidite":
            unite = "%"

        print(f"[INVALIDE] {type_capteur} {piece}: {valeur}{unite} ({raison})")
        channel.basic_nack(delivery_tag=method.delivery_tag, requeue=False)


def main():
    """Point d'entree : configure la connexion et lance la consommation."""
    connexion = connecter_amqp()
    canal = connexion.channel()

    # Declarer l'exchange "validated" de type topic
    canal.exchange_declare(exchange="validated", exchange_type="topic", durable=False)

    # Declarer une queue auto-generee et la lier a amq.topic avec le binding maison.#
    result = canal.queue_declare(queue="", exclusive=True)
    nom_queue = result.method.queue
    canal.queue_bind(exchange="amq.topic", queue=nom_queue, routing_key="maison.#")

    # Activer le manual ack (auto_ack=False)
    canal.basic_consume(queue=nom_queue, on_message_callback=callback, auto_ack=False)

    print("[AMQP] Service Validation demarre. En attente de messages...")
    print("[AMQP] Binding : amq.topic -> maison.#")
    print()

    try:
        canal.start_consuming()
    except KeyboardInterrupt:
        print("\n[AMQP] Arret du service Validation.")
        canal.stop_consuming()
        connexion.close()


if __name__ == "__main__":
    main()

File: session-3/nodejs/service_validation.js
#!/usr/bin/env node
/**
 * Service Validation — Verifie la coherence des donnees de capteurs.
 *
 * Ce service consomme les messages AMQP depuis l'exchange amq.topic (binding maison.#),
 * verifie que les valeurs sont dans les plages acceptables, puis :
 * - Si valide : republier sur l'exchange "validated" avec la meme routing key
 * - Si invalide : rejeter le message (nack, sans requeue)
 *
 * Utilise le manual ack pour garantir le traitement de chaque message.
 */

const amqplib = require("amqplib");

// Configuration AMQP
const AMQP_URL = "amqp://guest:guest@localhost:5672";

// Plages valides pour chaque type de capteur
const PLAGES_VALIDES = {
  temperature: { min: -20, max: 60 },
  humidite: { min: 0, max: 100 },
  fumee: { min: 0, max: 100 },
};

/**
 * Valide les donnees d'un capteur.
 * Retourne { valide: true } ou { valide: false, raison: "..." }
 */
function validerMessage(donnees) {
  const typeCapteur = donnees.sensor;
  const valeur = donnees.value;

  if (typeCapteur === "mouvement") {
    if (typeof valeur === "boolean") {
      return { valide: true };
    }
    return { valide: false, raison: `doit etre un booleen, recu: ${valeur}` };
  }

  const plage = PLAGES_VALIDES[typeCapteur];
  if (plage) {
    if (valeur >= plage.min && valeur <= plage.max) {
      return { valide: true };
    }
    return { valide: false, raison: `hors plage ${plage.min} a ${plage.max}` };
  }

  return { valide: false, raison: `type de capteur inconnu: ${typeCapteur}` };
}

/**
 * Connexion AMQP avec logique de retry (10 tentatives, 2s entre chaque).
 */
async function connecterAMQP() {
  for (let tentative = 1; tentative <= 10; tentative++) {
    try {
      const connexion = await amqplib.connect(AMQP_URL);
      console.log("[AMQP] Connecte a localhost");
      return connexion;
    } catch (err) {
      console.log(`[AMQP] Tentative ${tentative}/10 echouee : ${err.message}`);
      if (tentative < 10) {
        await new Promise((r) => setTimeout(r, 2000));
      }
    }
  }
  throw new Error("Connexion AMQP impossible apres 10 tentatives");
}

async function main() {
  const connexion = await connecterAMQP();
  const canal = await connexion.createChannel();

  // Declarer l'exchange "validated" de type topic
  await canal.assertExchange("validated", "topic", { durable: false });

  // Declarer une queue auto-generee et la lier a amq.topic
  const { queue: nomQueue } = await canal.assertQueue("", { exclusive: true });
  await canal.bindQueue(nomQueue, "amq.topic", "maison.#");

  console.log("[AMQP] Service Validation demarre. En attente de messages...");
  console.log("[AMQP] Binding : amq.topic -> maison.#");
  console.log();

  // Consommer les messages avec manual ack
  canal.consume(nomQueue, (msg) => {
    if (!msg) return;

    let donnees;
    try {
      donnees = JSON.parse(msg.content.toString());
    } catch (err) {
      console.log(`[INVALIDE] Message non-JSON recu: ${msg.content.toString()}`);
      canal.nack(msg, false, false); // nack sans requeue
      return;
    }

    const typeCapteur = donnees.sensor || "inconnu";
    const piece = donnees.room || "inconnue";
    const valeur = donnees.value;

    // Valider les donnees
    const resultat = validerMessage(donnees);

    if (resultat.valide) {
      // Republier sur l'exchange "validated" avec la meme routing key
      canal.publish("validated", msg.fields.routingKey, msg.content);
      console.log(`[VALIDE]   ${typeCapteur} ${piece}: ${valeur}`);
      canal.ack(msg);
    } else {
      // Rejeter le message sans requeue
      let unite = "";
      if (typeCapteur === "temperature") unite = "°C";
      else if (typeCapteur === "humidite") unite = "%";

      console.log(`[INVALIDE] ${typeCapteur} ${piece}: ${valeur}${unite} (${resultat.raison})`);
      canal.nack(msg, false, false);
    }
  }, { noAck: false });

  // Gestion de l'arret propre
  process.on("SIGINT", async () => {
    console.log("\n[AMQP] Arret du service Validation.");
    await canal.close();
    await connexion.close();
    process.exit(0);
  });
}

main().catch((err) => {
  console.error(`[ERREUR] ${err.message}`);
  process.exit(1);
});
